File: backend/app/recommendation_safety.py
""""Safety, clinical boundaries and explanation rules" — prompt 11 of the
nutrient-gap recommendation feature (see
docs/nutrient-gap-recommendations.md).

Centralises the rules every `recommend_*` module and the `/api/
recommendations/*` router must follow, so no module has to independently
re-decide them:

1. Never diagnose a nutrient deficiency — enforced by
   `nutrient_gap_analysis.NutrientStatus`'s vocabulary (below/near/within/
   above_preferred/above_upper_limit), never "deficient"/"excess".
2. Never claim a recommendation treats or prevents disease — enforced by
   wording convention across every `_explain()` in the `recommend_*`
   modules ("helps close the remaining X gap", never "treats X").
3. Never override a medically prescribed diet — `dietary_filter.py` never
   reads or acts on a `category="medical"` `DietaryConstraint` at all
   (see its module docstring); this module additionally surfaces
   `MEDICAL_CONSTRAINT_PRESENT` so a caller can display an explicit "this
   doesn't know your prescribed diet" notice rather than silent omission.
4. Never recommend a supplement — `candidate_metadata.py`'s curated
   table has no supplement entries and nothing in this feature queries
   a supplement category.
5. Never recommend exceeding a tolerable upper limit —
   `recommendation_scoring.py`'s `upper_limit_penalty` is large enough
   that `score_candidate` should never rank an upper-limit breach highly
   (see its adversarial regression tests); this module additionally
   tightens the upper limit itself for pregnancy/lactation (see
   `nutrient_targets.PREGNANCY_LACTATION_UPPER_LIMIT_MARGIN`).
6. Treat pregnancy, childhood, and other stored sensitive contexts
   conservatively — this module is where that's centralised (see
   `assess_eligibility` below).
7. Where target logic is insufficient for a sensitive profile, disable
   recommendations with a clear explanation rather than guessing — a
   profile below `MINIMUM_RECOMMENDATION_AGE` gets `enabled=False` here,
   since `energy_goal.py`/`protein_requirement.py`'s formulas are adult
   equations not validated for a growing child.
8/9/10. Estimates/variation/absorption disclaimers and the medical-advice
   boundary are the `SafetyWarningCode`s below, attached to every
   recommendation response rather than repeated in every card's prose.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum

from sqlalchemy.orm import Session

from .energy import calculate_age
from .models import DietaryConstraint, Profile

# Below this age, the app's personalised energy (Mifflin-St Jeor-style EER)
# and protein (per-kg factor) target formulas are standard adult
# equations — not validated for a growing child, whose requirements don't
# scale the same way and change quickly with development. Rather than
# feeding a wrong number into a recommendation and guessing, recommendations
# are disabled outright below this age (prompt 11 requirement 7).
MINIMUM_RECOMMENDATION_AGE = 18


class SafetyWarningCode(str, Enum):
    """Structured codes a caller (API consumer or frontend) can branch on
    without parsing prose — prompt 11's "structured warning codes rather
    than relying only on prose"."""

    DATA_IS_ESTIMATE = "data_is_estimate"
    RECIPE_NUTRIENTS_VARY = "recipe_nutrients_vary"
    ABSORPTION_VARIES = "absorption_varies"
    PREGNANCY_CONSERVATIVE = "pregnancy_conservative"
    LACTATION_CONSERVATIVE = "lactation_conservative"
    MEDICAL_CONSTRAINT_PRESENT = "medical_constraint_present"


WARNING_MESSAGES: dict[SafetyWarningCode, str] = {
    SafetyWarningCode.DATA_IS_ESTIMATE: (
        "Nutrient values come from reference food-composition data — actual "
        "content varies by brand, growing conditions and preparation."
    ),
    SafetyWarningCode.RECIPE_NUTRIENTS_VARY: (
        "A recipe's real nutrient content depends on the exact ingredients, "
        "brands and cooking method used, which can differ from what's shown here."
    ),
    SafetyWarningCode.ABSORPTION_VARIES: (
        "How much of a nutrient the body actually absorbs, and how much any "
        "individual needs, both vary — these are population reference values, "
        "not a personal measurement, and this is general nutritional "
        "information rather than medical advice."
    ),
    SafetyWarningCode.PREGNANCY_CONSERVATIVE: (
        "This profile is marked as pregnant — upper-limit comparisons are kept "
        "extra conservative here, but this remains general nutritional "
        "information, not antenatal medical advice."
    ),
    SafetyWarningCode.LACTATION_CONSERVATIVE: (
        "This profile is marked as lactating — upper-limit comparisons are kept "
        "extra conservative here, but this remains general nutritional "
        "information, not medical advice."
    ),
    SafetyWarningCode.MEDICAL_CONSTRAINT_PRESENT: (
        "This profile has a stored medical dietary consideration. This feature "
        "does not read that note and does not know your prescribed diet's "
        "specific requirements — it must not be used to override it. Check "
        "with whoever prescribed it before changing what you eat."
    ),
}


@dataclass(frozen=True)
class RecommendationEligibility:
    """Whether the recommendation engine should run at all for this
    profile, and which standing warnings apply regardless of outcome.
    `enabled=False` means the caller should show `disabled_reason` and
    skip calling suggest_ingredients/suggest_recipes/suggest_substitutions/
    suggest_pairs entirely — never a guess dressed up as a real answer."""

    enabled: bool
    disabled_reason: str | None
    warnings: list[SafetyWarningCode] = field(default_factory=list)


def assess_eligibility(profile: Profile, db: Session) -> RecommendationEligibility:
    age = calculate_age(profile)
    if age is not None and age < MINIMUM_RECOMMENDATION_AGE:
        return RecommendationEligibility(
            enabled=False,
            disabled_reason=(
                f"This profile is under {MINIMUM_RECOMMENDATION_AGE} — the energy and protein "
                "targets this feature compares against use adult formulas that aren't valid "
                "for a growing child, so recommendations are disabled here rather than "
                "guessing. A paediatric dietitian is the right source for a child's "
                "nutrition targets."
            ),
        )

    warnings = [SafetyWarningCode.DATA_IS_ESTIMATE, SafetyWarningCode.ABSORPTION_VARIES]
    if profile.is_pregnant:
        warnings.append(SafetyWarningCode.PREGNANCY_CONSERVATIVE)
    if profile.is_lactating:
        warnings.append(SafetyWarningCode.LACTATION_CONSERVATIVE)

    has_medical_constraint = (
        db.query(DietaryConstraint)
        .filter(DietaryConstraint.profile_id == profile.id, DietaryConstraint.category == "medical")
        .first()
        is not None
    )
    if has_medical_constraint:
        warnings.append(SafetyWarningCode.MEDICAL_CONSTRAINT_PRESENT)

    return RecommendationEligibility(enabled=True, disabled_reason=None, warnings=warnings)


def recipe_warnings(base: list[SafetyWarningCode]) -> list[SafetyWarningCode]:
    """Recipe-mode responses additionally warn that a recipe's real
    nutrient content depends on exact ingredients/preparation — appended
    to whatever profile-level warnings `assess_eligibility` already
    produced, never replacing them."""
    return [*base, SafetyWarningCode.RECIPE_NUTRIENTS_VARY]

File: frontend/src/lib/recommendationSafety.ts
// Mirrors app/recommendation_safety.py's SafetyWarningCode messages —
// prompt 11's "structured warning codes rather than relying only on
// prose". Shown once per panel (not repeated on every card) so a
// standing caveat like "these are estimates" or "this profile has a
// medical dietary consideration this feature can't read" is visible but
// unobtrusive, never omitted.
const WARNING_MESSAGES: Record<string, string> = {
	data_is_estimate:
		'Nutrient values come from reference food-composition data — actual content varies by brand, growing conditions and preparation.',
	recipe_nutrients_vary:
		"A recipe's real nutrient content depends on the exact ingredients, brands and cooking method used, which can differ from what's shown here.",
	absorption_varies:
		'How much of a nutrient the body actually absorbs, and how much any individual needs, both vary — these are population reference values, not a personal measurement, and this is general nutritional information rather than medical advice.',
	pregnancy_conservative:
		'This profile is marked as pregnant — upper-limit comparisons are kept extra conservative here, but this remains general nutritional information, not antenatal medical advice.',
	lactation_conservative:
		'This profile is marked as lactating — upper-limit comparisons are kept extra conservative here, but this remains general nutritional information, not medical advice.',
	medical_constraint_present:
		"This profile has a stored medical dietary consideration. This feature does not read that note and does not know your prescribed diet's specific requirements — it must not be used to override it. Check with whoever prescribed it before changing what you eat."
};

export function safetyWarningMessage(code: string): string {
	return WARNING_MESSAGES[code] ?? code;
}

// Mirrors app/recommend_ingredients.py's NoSuggestionReason — public-
// launch hardening prompt 4 item 6: "No safe or useful addition found"
// is a real, valid result, but a bare version of that message with no
// explanation is exactly the confusing dead end the prompt calls out.
const NO_SUGGESTION_REASON_MESSAGES: Record<string, string> = {
	no_shortfall: "Nothing tracked is currently below target for this — there's no gap to close.",
	no_eligible_candidates:
		'No practical, dietarily-suitable food in the catalogue stood out for this — try a different priority.',
	energy_limit:
		'Every food that would help was above the calorie limit set for this — try raising or removing it.',
	no_meaningful_improvement:
		"A few options were considered, but none would meaningfully move the needle — you're likely close to target already."
};

export function noSuggestionReasonMessage(code: string): string {
	return NO_SUGGESTION_REASON_MESSAGES[code] ?? 'No safe or useful addition found for the current priorities.';
}

// Operational-hardening prompt 3: labels/caveats for the carbon/
// glycaemic classification metadata in ScoreBreakdown (see
// RecommendationCard.svelte's "Why this ranked here" panel) — approximate,
// category-based preferences, never shown as a bare adjustment number
// with no tier/confidence/basis attached. Kept out of the component
// itself so the actual wording lives in one place, same convention as
// the safety/no-suggestion messages above.

const CARBON_TIER_LABELS: Record<string, string> = {
	very_high: 'very high',
	high: 'high',
	medium: 'medium',
	low: 'low'
};

export function carbonTierLabel(tier: string): string {
	return CARBON_TIER_LABELS[tier] ?? tier;
}

const GLYCAEMIC_TIER_LABELS: Record<string, string> = {
	high: 'high',
	medium: 'medium',
	low: 'low'
};

export function glycaemicTierLabel(tier: string): string {
	return GLYCAEMIC_TIER_LABELS[tier] ?? tier;
}

// "provenance" — whose name the classification actually came from.
// Recipes/pairs/substitutions have no single name of their own to check,
// so this is the honest "what this is actually based on" disclosure the
// prompt's "dominant-ingredient recipe classification can miss important
// ingredients" requirement calls for.
export function classificationProvenanceNote(provenance: 'name_match' | 'dominant_ingredient_proxy' | null): string {
	if (provenance === 'dominant_ingredient_proxy') {
		return "based on this recipe's largest-by-weight ingredient only — other ingredients aren't reflected, and a different one could change the result";
	}
	if (provenance === 'name_match') {
		return "based on this food's own name";
	}
	return '';
}

// "basis" — glycaemic-only: why a "low" classification landed there.
// Never let a negligible-carbohydrate food (meat, fish, eggs, cheese,
// nuts) read as if it had been lab-measured and scored low — GI simply
// doesn't apply to a food with negligible carbohydrate at all.
export function glycaemicBasisNote(basis: 'category_match' | 'negligible_carbohydrate' | null): string {
	if (basis === 'negligible_carbohydrate') {
		return "negligible carbohydrate — glycaemic index isn't really measured for foods like this, not a tested \"low\" result";
	}
	if (basis === 'category_match') {
		return 'a published category average, not this specific food measured';
	}
	return '';
}

export const CARBON_METHODOLOGY_NOTE =
	'A coarse, name-keyword estimate anchored to published food-carbon-footprint research — not a per-product lifecycle assessment, and not available for every food.';

export const GLYCAEMIC_METHODOLOGY_NOTE =
	"A coarse, name-keyword estimate, not a measured glycaemic index — GI itself isn't the same as glycaemic load (which also depends on portion size and carbohydrate quantity), and real GI varies with ripeness, variety and preparation. Common staples (bread, rice, potato, banana) are deliberately left unclassified rather than guessed. This is general nutritional information, not a prediction of your own blood-sugar response or medical advice.";
